[V1][Metrics] add support for kv event publishing

alec-flowers opened the pull request:
[RFC: KVBlocks and Metrics Publishing In Inference Frameworks](https://github.com/vllm-project/vllm/issues/16669)

- Added KVCacheEvent, BlockStored, BlockRemoved, and AllBlocksCleared msgspec classes
- Created a queue in the BlockPool and write these events in the appropriate functions
- Bubble the events up to the scheduler where they are appended to EngineCoreOutputs
- Add kv_cach_events to EngineCoreOutputs
- Wrote unit tests at the BlockManager level to test basic functionality and at the EngineCore level testing correct propagation and serializing over zmq.

API
- add enable_kv_cache_events to engineArgs
~- add external_stat_loggers field to AsyncLLM API~ Covered by https://github.com/vllm-project/vllm/pull/14661

With https://github.com/vllm-project/vllm/pull/14661 and this PR a 3rd party can write a custom stat logger to consume both engine Stats and Events and publish them elsewhere. 


<!--- pyml disable-next-line no-emphasis-as-heading -->

alec-flowers committed:
feat: add support for kv event publishing and custom statloggers

Signed-off-by: alec-flowers <aflowers@nvidia.com>

njhill commented:
@alec-flowers re the custom stats loggers, there's another PR for this which is hopefully close to being merged: https://github.com/vllm-project/vllm/pull/14661. As you alluded to, we need to take factories/constructors for the multi-engine case (or otherwise change the stats logger API to include an an engine id).

alec-flowers commented:
> @alec-flowers re the custom stats loggers, there's another PR for this which is hopefully close to being merged: #14661. As you alluded to, we need to take factories/constructors for the multi-engine case (or otherwise change the stats logger API to include an an engine id).

@njhill Awesome! Looks like there has been some good discussion around the API. I will remove my WIP stuff from this PR and reference that PR.

alec-flowers committed:
revert statlogger changes

Signed-off-by: alec-flowers <aflowers@nvidia.com>

robertgshaw2-redhat commented:
cc @markmc - FYI, this feature it to enable creating a global KV cache view in NVIDIA Dynamo

robertgshaw2-redhat commented on vllm/v1/engine/async_llm.py:
```diff
@@ -133,6 +134,8 @@ def from_vllm_config(
         # FIXME(rob): refactor VllmConfig to include the StatLoggers
         # include StatLogger in the Oracle decision.
+
+        # TODO (alec) ask rob about this
```
We were  not ready to support custom loggers since the apis were still in flux. @markmc should have final say about when we are ready for this

alec-flowers replied:
Yea this comment squeaked through from when I was thinking about adding it. The consensus is it should come from https://github.com/vllm-project/vllm/pull/14661, where @markmc has been involved.

alec-flowers replied:
I have removed.

robertgshaw2-redhat commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -684,6 +688,7 @@ def update_from_output(
         engine_core_outputs = EngineCoreOutputs(
             outputs=outputs,
             scheduler_stats=self.make_stats(spec_decoding_stats),
+            kv_cache_events=self.kv_cache_manager.get_kv_events(),
```
nit: I think using the method `take_events()` for consistency with the the `request.take_events()`

alec-flowers replied:
Yea I like that more. Its also an atomic read and then remove, so take fits better. I've propagated that to the BlockPool as well.

robertgshaw2-redhat commented:
This generally looks very reasonable to me.

It would be useful to add an example of the Logger which uses the the events, for testing purposes. Otherwise this will break.

Is there a reason why you don't want to implement the KVCacheEventsLogger inside VLLM?

alec-flowers committed:
address reviewer comments

- replace get_kv_events with take_events for consistency with request.take_events()
- remove wayward comment

Signed-off-by: alec-flowers <aflowers@nvidia.com>

alec-flowers commented:
> This generally looks very reasonable to me.
> 
> It would be useful to add an example of the Logger which uses the the events, for testing purposes. Otherwise this will break.
> 
> Is there a reason why you don't want to implement the KVCacheEventsLogger inside VLLM?

Main reason is to avoid adding dynamo (or pieces of dynamo) as a dependency to vLLM. For our logger, we would need to add in a rust based publisher (with python bindings) from Dynamo that sends the events over NATS to another component. 
https://github.com/ai-dynamo/dynamo/blob/main/lib/llm/src/kv_router/publisher.rs
https://github.com/ai-dynamo/dynamo/blob/main/lib/bindings/python/src/dynamo/_core.pyi#L497

I'm definitely flexible and open to ideas here. 

And good reminder, I will add in an integration test with a mock StatLogger that will break if the API changes too much.

ApostaC commented on vllm/v1/engine/__init__.py:
```diff
@@ -87,6 +87,31 @@ def new_event(cls,
         return cls(event_type, timestamp)
 
 
+class KVCacheEvent(
+        msgspec.Struct,
+        array_like=True,  # type: ignore[call-arg]
+        omit_defaults=True,  # type: ignore[call-arg]
+        gc=False,  # type: ignore[call-arg]
+        tag=True):
+    """Base class for all KV cache-related events"""
+
+
+class BlockStored(KVCacheEvent):
+    block_hashes: list[int]
+    parent_block_hash: Optional[int]
+    token_ids: list[int]
+    num_toks_per_block: list[int]
+    lora_id: Optional[int]
+
+
+class BlockRemoved(KVCacheEvent):
+    block_hashes: list[int]
+
+
+class AllBlocksCleared(KVCacheEvent):
+    pass
```
Should those events be defined in a source file that is more related to KV cache management? For example, `kv_cache_utils.py`

ApostaC commented on vllm/v1/core/kv_cache_manager.py:
```diff
@@ -383,3 +386,11 @@ def free_block_hashes(self, request: Request) -> None:
         """
         self.req_to_block_hashes.pop(request.request_id, None)
+
+    def take_events(self) -> list[KVCacheEvent]:
```
nit: `take_events` -> `get_kv_events`

ApostaC reviewed (commented):
Quick question: how is `KVCacheEvent` published? I saw it's passed into the metrics logger, but seems like it's not used?

alec-flowers commented:
> Quick question: how is `KVCacheEvent` published? I saw it's passed into the metrics logger, but seems like it's not used?

It would be published by a 3rd party defined Logger. You would be free to write your own that would publish it how you want. That is why in this case we would be reliant on this issue - https://github.com/vllm-project/vllm/pull/14661, which adds support for passing in a logger at runtime to the AysncLLM class.

alec-flowers committed:
remove kv_cache_events from EngineCoreOutputs

Signed-off-by: alec-flowers <aflowers@nvidia.com>

alec-flowers committed:
add zmq publisher for kv cache events

Signed-off-by: alec-flowers <aflowers@nvidia.com>

alec-flowers commented:
@markmc I've changed the publishing to hook into a similar point as the KV Connector API. It follows the though process I laid out in the RFC. 
https://github.com/vllm-project/vllm/issues/16669#issuecomment-2819245423

markmc commented:
> @markmc I've changed the publishing to hook into a similar point as the KV Connector API. It follows the thought process I laid out in the RFC. [#16669 (comment)](https://github.com/vllm-project/vllm/issues/16669#issuecomment-2819245423)

This all looks really excellent, Alec.

I really only have a high level comment - KV cache events (distributed state) looks unrelated to "metrics" (quantitative insights?) , but looks very much in scope for the `vllm/distributed` package and would live nicely alongside `kv_transfer`

(I know this isn't quite the same as what I said before - "the new KV connector API seems like a much closer fit" - but that's just my own ignorance!)

So, I would imagine moving all of the following to `vllm.distributed.kv_events`

- `vllm.v1.metrics.events`
- `vllm.v1.core.kv_cache_utils.KVCacheEvent` etc.

and renaming `EventPublisherConfig` to `KVEventsConfig`, also moving `enable_kv_cache_events` to `KVEventsConfig`

wdyt?

alec-flowers commented:
Yea, I definitely agree. And coupling the enable_kv_events to the publishing is also the right way to go. I will make those changes right away. 

> > @markmc I've changed the publishing to hook into a similar point as the KV Connector API. It follows the thought process I laid out in the RFC. [#16669 (comment)](https://github.com/vllm-project/vllm/issues/16669#issuecomment-2819245423)
> 
> This all looks really excellent, Alec.
> 
> I really only have a high level comment - KV cache events (distributed state) looks unrelated to "metrics" (quantitative insights?) , but looks very much in scope for the `vllm/distributed` package and would live nicely alongside `kv_transfer`
> 
> (I know this isn't quite the same as what I said before - "the new KV connector API seems like a much closer fit" - but that's just my own ignorance!)
> 
> So, I would imagine moving all of the following to `vllm.distributed.kv_events`
> 
> * `vllm.v1.metrics.events`
> * `vllm.v1.core.kv_cache_utils.KVCacheEvent` etc.
> 
> and renaming `EventPublisherConfig` to `KVEventsConfig`, also moving `enable_kv_cache_events` to `KVEventsConfig`
> 
> wdyt?

Yea, I definitely agree. And coupling the enable_kv_events to the publishing is also the right way to go. I will make those changes right away.

alec-flowers committed:
re-arrange files

Signed-off-by: alec-flowers <aflowers@nvidia.com>

alec-flowers commented:
@markmc I've gone ahead and made those changes!

alec-flowers commented:
I'm passing most of the initial checks. However the docker build seems to be failing. It doesn't look related to any of my changes as I don't edit the docker nor change dependencies. 
Any thoughts @markmc @robertgshaw2-redhat

markmc committed:
[V1][KV Events] Add example script and publisher

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -723,7 +738,8 @@ def update_from_output(
                         stop_reason=request.stop_reason,
-                        events=request.take_events()))
+                        events=request.take_events(),
+                    ))
```
Revert (but only if you're making other changes)

alec-flowers replied:
Done.

markmc reviewed (commented):
A couple of very minor suggestions - assume fixed `block_size`, and make `publish()` construct the batch object

markmc commented:
I played around with this a little and added an example subscriber. Feel free to pull in if you think it's good to include :+1: 

```
$ git fetch https://github.com/markmc/vllm.git alec-flowers-kv-event-publishing
$ git merge --ff-only FETCH_HEAD
```

alec-flowers commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -492,6 +502,11 @@ def schedule(self) -> SchedulerOutput:
+        events = self.kv_cache_manager.take_events()
+        if events:
+            batch = KVEventBatch(ts=time.time(), events=events)
+            self.kv_event_publisher.publish(batch)
```
The reason I did it this way is the publisher is written to take general events. So we could potentially re-use it elsewhere. If we make publisher create KVEventBatch then it can only be used here.

alec-flowers replied:
And thanks for the time suggestion! I've never used monotonic() before. I will make the change.

alec-flowers replied:
comparable between processes assuming synced time
Yes. I think I should have actually considered it rather than just made the change. Given that we are going to be in a distributed system with different processes we need to refer to a similar time, which is what time.time() gives us.

alec-flowers commented on vllm/v1/engine/async_llm.py:
```diff
@@ -380,8 +381,7 @@ async def output_handler():
                             scheduler_stats=outputs.scheduler_stats,
-                            iteration_stats=iteration_stats,
-                        )
+                            iteration_stats=iteration_stats)
```
Done

alec-flowers commented on vllm/v1/engine/async_llm.py:
```diff
@@ -74,6 +74,7 @@ def __init__(
                     loggers.append(LoggingStatLogger(engine_index=i))
                 loggers.append(
                     PrometheusStatLogger(vllm_config, engine_index=i))
+
```
Done

alec-flowers committed:
change to block_size

Signed-off-by: alec-flowers <aflowers@nvidia.com>

alec-flowers committed:
remove lingering spaces, adjust time to monotonic

Signed-off-by: alec-flowers <aflowers@nvidia.com>

alec-flowers committed:
add dynamic dir so can run bash command from anywhere

Signed-off-by: alec-flowers <aflowers@nvidia.com>

alec-flowers commented:
> I played around with this a little and added an example subscriber. Feel free to pull in if you think it's good to include 👍
> 
> ```
> $ git fetch https://github.com/markmc/vllm.git alec-flowers-kv-event-publishing
> $ git merge --ff-only FETCH_HEAD
> ```

Thats awesome! Thanks for adding. I've pulled it in.

alec-flowers commented:
@njhill @WoosukKwon @ywang96 Would I be able to get a code-owner review?

njhill commented on vllm/distributed/kv_events.py:
```diff
@@ -0,0 +1,196 @@
+        for seq, buf in self._buffer:
+            if seq >= start_seq:
+                # [identity, empty_delim, seq_bytes, payload]
+                self._replay.send_multipart(
+                    [client_id, b"",
+                     seq.to_bytes(8, "big"), buf])
+        # Send end of sequence marker
+        self._replay.send_multipart([client_id, b"", b""])
```
nit:
```suggestion
                self._replay.send_multipart(
                    (client_id, b"",
                     seq.to_bytes(8, "big"), buf))
        # Send end of sequence marker
        self._replay.send_multipart((client_id, b"", b""))
```

alec-flowers replied:
Added

njhill reviewed (commented):
Thanks @alec-flowers the code looks very nice, and really nice tests.

Another consideration is data parallel deployments, where there will be multiple engines (with their own schedulers + kvcaches) deployed as a single vLLM instance. These are configured via a single VllmConfig but would likely need separate queues for publishing kv cache events.

alec-flowers commented:
> Thanks @alec-flowers the code looks very nice, and really nice tests.
> 
> Another consideration is data parallel deployments, where there will be multiple engines (with their own schedulers + kvcaches) deployed as a single vLLM instance. These are configured via a single VllmConfig but would likely need separate queues for publishing kv cache events.

If they have their own scheduler, then they should have their own queue. Having the same VllmConfig might be an issue as we would need to adjust the zmq topology and ensure we use the right bind/connect things. 
For data parallel that is a case that Dynamo works well for and is why we are adding this support here.

alec-flowers commented on vllm/distributed/kv_events.py:
```diff
@@ -0,0 +1,196 @@
+            self._replay = self._ctx.socket(zmq.ROUTER)
+            self._replay.bind(replay_endpoint)
+
+        self._buffer: collections.deque[tuple[int,
+                                              bytes]] = (collections.deque(
+                                                  maxlen=buffer_steps))
```
Nice suggestion! Didn't even know about this typing syntax - deque[tuple[int, bytes]]

alec-flowers commented on vllm/distributed/kv_events.py:
```diff
@@ -0,0 +1,196 @@
+        if not self._replay.poll(0):  # no request waiting
+            return
+        # Receive request frames: [client_id, start_seq_bytes]
+        frames = self._replay.recv_multipart()
+        client_id = frames[0]
+        start_seq = int.from_bytes(frames[-1], "big")
```
Added!

alec-flowers commented on vllm/v1/core/block_pool.py:
```diff
@@ -206,6 +232,9 @@ def _maybe_evict_cached_block(self, block: KVCacheBlock) -> bool:
 
+            if self.enable_kv_cache_events:
+                self.kv_event_queue.append(
+                    BlockRemoved(block_hashes=[block_hash.hash_value], ))
```
Changed

alec-flowers commented on vllm/distributed/kv_events.py:
```diff
@@ -0,0 +1,196 @@
+        payload = self._pack.encode(events)
+        seq_bytes = self._seq.to_bytes(8, "big")
+
+        self._pub.send_multipart([self._topic_bytes, seq_bytes, payload])
```
Changed

alec-flowers commented on vllm/v1/core/block_pool.py:
```diff
@@ -116,6 +126,8 @@ def cache_full_blocks(
             prev_block_hash_value = prev_block.block_hash.hash_value
 
+        parent_block_hash = prev_block_hash_value
+        new_hashes: list[int] = []
```
Added

alec-flowers commented on vllm/v1/core/block_pool.py:
```diff
@@ -153,8 +165,22 @@ def cache_full_blocks(
             # Update and added the full block to the cache.
             blk.block_hash = block_hash
             self.cached_block_hash_to_block[block_hash][blk.block_id] = blk
+            new_hashes.append(block_hash.hash_value)
```
Added

alec-flowers commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -62,6 +64,9 @@ def __init__(
             self.scheduler_config.max_num_batched_tokens
         self.max_model_len = self.scheduler_config.max_model_len
+        self.enable_kv_cache_events = (
+            self.kv_events_config.enable_kv_cache_events
+            if self.kv_events_config is not None else False)
```
Added

alec-flowers commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -492,6 +502,11 @@ def schedule(self) -> SchedulerOutput:
+        events = self.kv_cache_manager.take_events()
+        if events:
+            batch = KVEventBatch(ts=time.monotonic(), events=events)
+            self.kv_event_publisher.publish(batch)
```
Yea I was actually leaning towards this because this scheduler is single threaded. So even though I moved it out of the main Outputs path as you note its still on the critical path.

njhill replied:
I think this could all be encapsulated in the ZmqEventPublisher btw

njhill replied:
And sorry I know this might not be relevant to you since you have your own publisher

alec-flowers replied:
No problem! I've added a separate thread. Let me know what you think.

alec-flowers commented on examples/online_serving/kv_events_subscriber.py:
```diff
@@ -0,0 +1,67 @@
+while True:
+    try:
+        _, seq_bytes, payload = socket.recv_multipart()
+        seq = int.from_bytes(seq_bytes, "big")
```
Yes, just added.

alec-flowers committed:
reviewer fixes

Signed-off-by: alec-flowers <aflowers@nvidia.com>

alec-flowers committed:
add replays to kv event subscriber example

Signed-off-by: alec-flowers <aflowers@nvidia.com>

alec-flowers committed:
use thread to publish events

Signed-off-by: alec-flowers <aflowers@nvidia.com>

njhill commented on vllm/v1/core/sched/interface.py:
```diff
@@ -137,3 +137,8 @@ def make_stats(self) -> Optional["SchedulerStats"]:
+    @abstractmethod
+    def shutdown(self) -> None:
+        """Shutdown the scheduler."""
+        raise NotImplementedError
```
nit: trailing newline

alec-flowers replied:
Added

njhill commented on vllm/distributed/kv_events.py:
```diff
@@ -0,0 +1,290 @@
+                self._replay.send_multipart(
+                    (client_id, b"", seq.to_bytes(8, "big"), buf))
+        # Send end of sequence marker
+        self._replay.send_multipart((client_id, b"", b""))
```
Shouldn't this have four elements?

alec-flowers replied:
This is a bit of a judgement call how you want to send an end of sequence marker. The cliend_id and first b"" are actually stripped off by the router and its just the (seq, buf) that is sent. So the same happens here and I was choosing to just send a single element of empty bytes as the sequence marker. 

However after thinking about it more, I think I would prefer to keep the recv interface to have a constant size. At the cost of a few more bytes I changed it now so that they are both 4 elements. I think this makes the handling a bit more easy and intuitive.

njhill commented on examples/online_serving/kv_events_subscriber.py:
```diff
@@ -0,0 +1,111 @@
+                    while poller.poll(timeout=2000):
+                        _, delimiter, seq_bytes, replay_payload = (
+                            replay.recv_multipart())
+                        if not delimiter:
```
I think delimiter will always be false? (empty bytes)

alec-flowers replied:
Yea this is just wrong. I handle it right in the tests but forgot the correct interface here... The path isn't actually hit in the small test we run. I fixed it.

njhill commented on examples/online_serving/kv_events_subscriber.py:
```diff
@@ -0,0 +1,111 @@
+
+    while True:
+        try:
+            if sub.poll(50):
```
I don't think the timeout/if is needed?
```suggestion
            sub.poll():
```

alec-flowers replied:
Technically we can do that. However this code in my head is more of an example of how to use it, and likely you would have other code you might want to run after the timeout but before the loop cycles around. The if statement allows you to do that vs. hitting a blocking sub.poll().

alec-flowers replied:
I added a code comment to note this.

njhill reviewed (commented):
Thanks @alec-flowers, looks great

alec-flowers committed:
address reviewer comments

Signed-off-by: alec-flowers <aflowers@nvidia.com>

alec-flowers committed:
Merge branch 'main' into kv-event-publishing

Signed-off-by: alec-flowers <aflowers@nvidia.com>

njhill commented on examples/online_serving/kv_events_subscriber.py:
```diff
@@ -0,0 +1,114 @@
+                    replay.send((last_seq + 1).to_bytes(8, "big"))
+
+                    while poller.poll(timeout=200):
+                        seq_bytes, replay_payload = (replay.recv_multipart())
```
nit:
```suggestion
                        seq_bytes, replay_payload = replay.recv_multipart()
```

njhill commented on tests/distributed/conftest.py:
```diff
@@ -0,0 +1,145 @@
+                    break
+
+                frames = self.replay.recv_multipart()
+                if not frames or (not frames[-1]):
```
nit:
```suggestion
                if not frames or not frames[-1]:
```

njhill reviewed (approved):
Thanks @alec-flowers!

njhill commented:
@alec-flowers it looks like there may be an issue with the `test_kv_cache_events` test timing out: https://buildkite.com/vllm/ci/builds/18784#01967c2d-91f5-42bb-b3ae-ce520c3ddb11/210-1613. It might help to set `daemon=True` when starting the publishing thread.

alec-flowers committed:
fixes, shutdown to kv_event_test

Signed-off-by: alec-flowers <aflowers@nvidia.com>

alec-flowers committed:
fix: attempt to fix hanging test

Signed-off-by: alec-flowers <aflowers@nvidia.com>

alec-flowers committed:
fix: leaked thread from test_engine_core_client_asyncio

Signed-off-by: alec-flowers <aflowers@nvidia.com>

alec-flowers committed:
Merge branch 'main' into kv-event-publishing

Signed-off-by: alec-flowers <aflowers@nvidia.com>

alec-flowers committed:
Merge branch 'main' into kv-event-publishing

Signed-off-by: alec-flowers <aflowers@nvidia.com>

alec-flowers commented:
Took me some time, but isolated the problem to be the previous test_engine_core_client_asyncio which was leaving a thread open with a zmq socket that was causing problems with my test. I believe I've fixed it.

@njhill I'm having some problems with CI. It seems the last run failed in the build stage. I'm also constantly having to re-base due to a large import block. Hopefully this run goes smooth.

njhill commented:
Thanks @alec-flowers, sorry you had to take extra time to debug that. I think it's a lingering issue where the client destructor doesn't always get run. In theory the explicit shutdown shouldn't be needed but makes sense to include it here for now.
